### PR TITLE
#45641 - Adjusted Resource Limit Logic

### PR DESCRIPTION
**Pull request by gehrkefc**
## Problem
The user wasn't able to set Limits or Requests for namespace when it wasn't previously defined in project. 
 
## Solution
We changed the logic to the user be able to set the values in namespace and when some value is missing, we copy it from project, this way project values work like default values.
 
## Testing
The repro steps are in the original issue. For the testing, I suggest you to create a new project in the UI and then set some values for Limits or Requests. Then after, create a namespace in that project and add other values. The values you add in namespace will be the ones set, but in case you don't set one or more of them, it will copy it from project if you previously defined it in there.

### CODE REVIEW

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go [hunk 1]
```diff
@@ -4,14 +4,169 @@ import (
...
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/api/resource"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
 )
 
+func TestCompleteLimit(t *testing.T) {
+	type input struct {
+		newValues     *v32.ContainerResourceLimit
+		defaultValues *v32.ContainerResourceLimit
```
Comment (tomleb) [inline pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go:18]: Might be more readable like this:

```suggestion
		nsLimits     *v32.ContainerResourceLimit
		projectLimits *v32.ContainerResourceLimit
```

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go [hunk 2]
```diff
@@ -4,14 +4,169 @@ import (
...
+	type expected struct {
+		expected *v32.ContainerResourceLimit
+		err      error
+	}
+
+	testCases := []struct {
+		name     string
+		input    input
+		expected expected
+	}{
```
Comment (tomleb) [inline pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go:30]: One more case that could be tested is project limit being nil.

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync.go [hunk 3]
```diff
@@ -456,31 +457,34 @@ func completeQuota(requestedQuota *v32.ResourceQuotaLimit, defaultQuota *v32.Res
 	return toReturn, err
 }
 
-func completeLimit(existingLimit *v32.ContainerResourceLimit, defaultLimit *v32.ContainerResourceLimit) (*v32.ContainerResourceLimit, error) {
+func completeLimit(newLimit *v32.ContainerResourceLimit, defaultLimit *v32.ContainerResourceLimit) (*v32.ContainerResourceLimit, error) {
```
Comment (tomleb) [inline pkg/controllers/managementuser/resourcequota/resource_quota_sync.go:460]: Would probably be even less confusing this way

```suggestion
func completeLimit(nsLimit *v32.ContainerResourceLimit, projectLimit *v32.ContainerResourceLimit) (*v32.ContainerResourceLimit, error) {
```

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync.go [hunk 4]
```diff
@@ -7,19 +7,20 @@ import (
...
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime"
 	clientcache "k8s.io/client-go/tools/cache"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	v1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	namespaceutil "github.com/rancher/rancher/pkg/namespace"
+	validate "github.com/rancher/rancher/pkg/resourcequota"
+	"github.com/rancher/rancher/pkg/utils"
```
Comment (tomleb) [inline pkg/controllers/managementuser/resourcequota/resource_quota_sync.go:23]: nit: I don't think this change is necessary, afaik other files don't separate rancher-specific imports vs others.

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go [hunk 5]
```diff
@@ -4,14 +4,169 @@ import (
...
 	"testing"
 
-	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
-
 	"github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
 	apiequality "k8s.io/apimachinery/pkg/api/equality"
 	"k8s.io/apimachinery/pkg/api/resource"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
```
Comment (tomleb) [inline pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go:12]: nit: I don't think this change is necessary, afaik other files don't separate rancher-specific imports vs others.

**Review by tomleb — COMMENTED** [review]
woops I just saw you already had 2 reviewers on this after writting my comments.. I'll send them either way since I already spent the time on it..

Overall LGTM and thanks for the tests

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go (hunk 1 above)
Comment (gehrkefc) [reply pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go]: Ok, Thanks!

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go (hunk 2 above)
Comment (gehrkefc) [reply pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go:30]: Sure, just added. Thanks!

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync.go (hunk 3 above)
Comment (gehrkefc) [reply pkg/controllers/managementuser/resourcequota/resource_quota_sync.go]: Sure, just changed it. Thanks!

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync.go (hunk 4 above)
Comment (gehrkefc) [reply pkg/controllers/managementuser/resourcequota/resource_quota_sync.go]: Just reverted it. Thanks!

File: pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go (hunk 5 above)
Comment (gehrkefc) [reply pkg/controllers/managementuser/resourcequota/resource_quota_sync_test.go]: Just fixed it. Thanks!

**Review by tomleb — APPROVED** [review]
